### ISSUE TITLE
Stricter matching rule to support demo and testing 

### ISSUE THREAD

**Issue by khanh-nguyen** [jamesdbloom/mockserver#144]
When demo, I just want to have a simple rule like this:
```
{
    httpRequest: {
        method: 'GET',
        path: '/test/endpoint'
    },
    httpResponse: {
        statusCode: 200
    },
    times: {
    		unlimited: true
    }
}
```
However, for parallel testing, I use expectations like this:
```
{
    httpRequest: {
        method: 'GET',
        path: '/test/endpoint',
        cookies: [
						{
								name: 'some name',
								value: 'some value'
						}
				]
    },
    httpResponse: {
        statusCode: 200
    },
    times: {
    		unlimited: true
    }
}
```
I want to have both rules in place so we can go on with the demo, but still be able to run tests. The problem is that the first rule is very flexible, that the second rule becomes in effective. If the first rule is in place, then we can query `/test/endpoint` with `cookie="whatever name=whatever value"` and still get back `200`.

Is it possible to make a **stricter** matching rule to have both rules as above, @jamesdbloom ?

**Comment by jamesdbloom** [comment]
The MockServer does support the ability to `not` a match.  Therefore you could add to the first expectation `not` cookie as follows:

```js
{
    "httpRequest": {
        "method": "GET", 
        "path": "/cookieNot", 
        "cookies": [
            {
                "not": true,
                "name": "name", 
                "value": "value"
            }
        ]
    }, 
    "httpResponse": {
        "statusCode": 200
    }, 
    "times": {
        "unlimited": true
    }
}
```

However as this is a new feature is currently only works if the cookie name does not match if just the cookie value does not match then the `not` does not work.  I can fix this as it is a bug.

**Comment by khanh-nguyen** [comment]
Cool. I'll try that out. However, `value` should be generated randomly in order to support parallel tests. So, I'd expect some thing like this to work:
```
"cookies": [
            {
                "not": true,
                "name": "name", 
                "value": ".*"
            }
        ]
```
or even better
```
"cookies": [
            {
                "not": true,
                "name": "name"
            }
        ]
```
, right?

**Comment by jamesdbloom** [comment]
Yes the following should work currently, however this only `not`s the `name` and has no impact of the matching of the `value`, but I think this is what you need anyway:

```
"cookies": [
            {
                "not": true,
                "name": "name", 
                "value": ".*"
            }
        ]
```

The `not` logic was fairly recently implemented and has not been fully applied to Cookies, Parameter and Headers.  Now I been going through it and added extra tests I realised there are some corner cases that are almost impossible to solve with code.  So I'm going to have to re-implement it so that it will be possible to `not` either the `name` or `value`, using `!`, as follows:

```
"cookies": [
            {
                "name": "!name", 
                "value": ".*"
            }
        ]
```

Once I get this completed I will release all the recent fixes.

**Comment by jamesdbloom** [comment]
I am still working on this, it definitely turned out that fixing the `not` functionality in a consistent and clear way for cookiers, headers and parameters is more involved then I have hopped.   

So far I have added > 400 additional tests to make sure I have covered all the scenarios I need to do and that is only when I cleaned up the NottableRegexMultiMap and NottableRegexHashMap implementations.

Anyway hopefully in the next week I should have it completed as I've done the most laborious parts now.
